### PR TITLE
__restrict in arithmetic functions

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
More strict aliasing for inner loops of arithmetic functions.

### CODE REVIEW

File: dbms/src/Functions/FunctionBinaryArithmetic.h [hunk 1]
```diff
@@ -63,23 +63,20 @@ struct BinaryOperationImplBase
...
-    static void NO_INLINE vector_constant(const PaddedPODArray<A> & a, B b, PaddedPODArray<ResultType> & c)
+    static void NO_INLINE vector_constant(const A * __restrict a, B b, ResultType * __restrict c, size_t size)
     {
-        size_t size = a.size();
         for (size_t i = 0; i < size; ++i)
             c[i] = Op::template apply<ResultType>(a[i], b);
     }
 
-    static void NO_INLINE constant_vector(A a, const PaddedPODArray<B> & b, PaddedPODArray<ResultType> & c)
+    static void NO_INLINE constant_vector(A a, const B * __restrict b, ResultType * __restrict c, size_t size)
```
Comment (amosbird) [inline dbms/src/Functions/FunctionBinaryArithmetic.h]: Does it actually generate different code? If so, does it mean that PODArray needs optimization?

Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionBinaryArithmetic.h]: I did not checked it yet.
I expect that if B and C are the same types, aliasing will prevent vectorization.

File: dbms/src/Functions/intDiv.cpp [hunk 2]
```diff
@@ -21,7 +21,7 @@ struct DivideIntegralByConstantImpl
     using ResultType = typename DivideIntegralImpl<A, B>::ResultType;
     static const constexpr bool allow_fixed_string = false;
 
-    static void vector_constant(const PaddedPODArray<A> & a, B b, PaddedPODArray<ResultType> & c)
+    static void vector_constant(const A * __restrict a_pos, B b, ResultType * __restrict c_pos, size_t size)
```
Comment (amosbird) [inline dbms/src/Functions/intDiv.cpp]: NO_INLINE, I'm not sure if GCC handles inline and __restrict correctly but vectorized function without inlining sounds reasonable.

File: dbms/src/Functions/modulo.cpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ struct ModuloByConstantImpl
     using ResultType = typename ModuloImpl<A, B>::ResultType;
     static const constexpr bool allow_fixed_string = false;
 
-    static void vector_constant(const PaddedPODArray<A> & a, B b, PaddedPODArray<ResultType> & c)
+    static void vector_constant(const A * __restrict src, B b, ResultType * __restrict dst, size_t size)
```
Comment (amosbird) [inline dbms/src/Functions/modulo.cpp]: NO_INLINE

File: dbms/src/Functions/FunctionBinaryArithmetic.h (hunk 1 above)
Comment (amosbird) [reply dbms/src/Functions/FunctionBinaryArithmetic.h]: Ok. Hope we could observe promising results :)

**Comment by alexey-milovidov** [comment]
Nothing changed according to performance tests.

**Comment by alexey-milovidov** [comment]
Yes, it makes sense. See https://github.com/ClickHouse/ClickHouse/pull/9091#issuecomment-590421657 (this was the motivation for this change)

**Pull request merged**